src/trace_writer.ml

```diff
@@ -655,6 +636,109 @@ let check_current_symbol
+    for _ = 0 to !frames_to_unwind + diff do
+      ret t thread_info ~time
+    done;
+    frames_to_unwind := 0
+  ;;
+
+  let ret_track_exn_data t thread_info ~time =
+    let { Thread_info.callstack; ocaml_exception_state; _ } = thread_info in
+    (match ocaml_exception_state with
+    | With_exception_info _ -> ()
+    | Without_exception_info { frames_to_unwind } ->
+      (match Callstack.top callstack with
+      | Some (From_perf symbol) ->
+        (match symbol with
+        | "caml_next_frame_descriptor" -> incr frames_to_unwind
+        | "caml_raise_exn" -> unwind_stack t thread_info ~time ~frames_to_unwind (-2)
+        | "caml_raise_exception" -> unwind_stack t thread_info ~time ~frames_to_unwind 1
+        | _ -> ())
+      | _ -> ()));
+    ret t thread_info ~time
+  ;;
+
+  let clear_trap_stack t thread_info ~time =
+    clear_callstack t thread_info ~time;
+    match Stack.pop thread_info.inactive_callstacks with
+    | Some callstack -> thread_info.callstack <- callstack
+    | None -> thread_info.callstack <- Callstack.create ~create_time:time
+  ;;
+
+  let check_current_symbol_track_entertraps
+      t
+      (thread_info : 'a Thread_info.t)
+      ~time
+      (dst : Event.Location.t)
+    =
+    match thread_info.ocaml_exception_state with
+    | With_exception_info { ocaml_exception_info; _ }
+      when Ocaml_exception_info.is_entertrap
+             ocaml_exception_info
+             ~addr:dst.instruction_pointer -> clear_trap_stack t thread_info ~time
+    | _ -> check_current_symbol t thread_info ~time dst
+  ;;
+
+  let track_executed_pushtraps_and_poptraps_in_range
+      t
+      (thread_info : _ Thread_info.t)
+      ~(src : Event.Location.t)
+      ~(dst : Event.Location.t)
+      ~time
+    =
+    match thread_info.ocaml_exception_state with
+    | Without_exception_info _ -> ()
+    | With_exception_info { ocaml_exception_info; last_known_instruction_pointer } ->
+      (match !last_known_instruction_pointer with
+      | None -> ()
+      | Some last_known_instruction_pointer ->
+        Ocaml_exception_info.iter_pushtraps_and_poptraps_in_range
+          ocaml_exception_info
+          ~from:last_known_instruction_pointer
+          ~to_:src.instruction_pointer
+          ~f:(fun (addr, kind) ->
+            match kind with
+            | Poptrap -> clear_trap_stack t thread_info ~time
+            | Pushtrap ->
+              Stack.push thread_info.inactive_callstacks thread_info.callstack;
+              thread_info.callstack <- Callstack.create ~create_time:time;
+              if !debug
+              then
+                call
+                  t
+                  thread_info
+                  ~time
+                  ~location:
+                    { instruction_pointer = 0L
+                    ; symbol_offset = 0
+                    ; symbol = From_perf [%string "[push trap @ %{addr#Int64.Hex}]"]
+                    }));
+      last_known_instruction_pointer := Some dst.instruction_pointer
```

Comment:
Why did you choose a ref for last_known_instruction_pointer instead of using a mutable field?

Reply:
We were already doing this for `frames_to_unwind` and I didn't want to change that code in this PR.